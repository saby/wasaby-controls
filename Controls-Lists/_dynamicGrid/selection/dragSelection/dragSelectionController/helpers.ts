import { TPoint, TRect } from './types';
import { TResizerDirection } from '../shared/types';
import { START_FIXED_PART_OBSERVER, END_FIXED_PART_OBSERVER } from '../../../shared/constants';
import { TKeyPair } from './../../shared/types';
import * as React from 'react';

const PERCENT_OF_SIZE_TO_SELECT_CELL = 0.1;

export function getDragStartParams(
    container: HTMLElement,
    topLeftRect: TRect,
    bottomRightRect: TRect,
    direction: TResizerDirection
): {
    startPoint: TPoint;
    xOffset: number;
    yOffset: number;
    xMinOffset: number;
    yMinOffset: number;
} {
    const viewPortRect = getViewportBoundingClientRect(container);
    let xOffset: number;
    let yOffset: number;
    let xMinOffset: number;
    let yMinOffset: number;

    const startPoint: TPoint = {
        x: undefined,
        y: undefined,
    };

    const getXInViewportRange = (x: number) =>
        Math.min(Math.max(x, viewPortRect.x), viewPortRect.x + viewPortRect.width);
    const getYInViewportRange = (y: number) =>
        Math.min(Math.max(y, viewPortRect.y), viewPortRect.y + viewPortRect.height);

    const getXOffsetInViewportRange = (startX: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartX = getXInViewportRange(startX + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartX - startX : -1 * (startX - correctStartX);
    };

    const getYOffsetInViewportRange = (startY: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartY = getYInViewportRange(startY + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartY - startY : -1 * (startY - correctStartY);
    };

    // Координата Y
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(bottomRightRect.y + bottomRightRect.height);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            yOffset = getYOffsetInViewportRange(startPoint.y, topLeftRect.y - startPoint.y);
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(topLeftRect.y);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            yOffset = getYOffsetInViewportRange(
                startPoint.y,
                bottomRightRect.y + bottomRightRect.height - startPoint.y
            );
            break;
    }

    // Координата X
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(bottomRightRect.x + bottomRightRect.width);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            xOffset = getXOffsetInViewportRange(startPoint.x, topLeftRect.x - startPoint.x);
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(topLeftRect.x);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            xOffset = getXOffsetInViewportRange(
                startPoint.x,
                bottomRightRect.x + bottomRightRect.width - startPoint.x
            );
            break;
    }

    // Минимальное смещение по Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            yMinOffset = -1 * bottomRightRect.height;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            yMinOffset = topLeftRect.height;
            break;
        case 'left':
        case 'right':
            yMinOffset = yOffset;
            break;
    }

    // Минимальное смещение по X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            xMinOffset = -1 * bottomRightRect.width;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            xMinOffset = topLeftRect.width;
            break;
        case 'top':
        case 'bottom':
            xMinOffset = xOffset;
            break;
    }

    return {
        startPoint,
        xOffset,
        yOffset,
        xMinOffset,
        yMinOffset,
    };
}

export function getNewOffset(
    cellRect: TRect,
    startPoint: TPoint,
    direction: TResizerDirection,
    minXOffset: number,
    minYOffset: number
): {
    x: number;
    y: number;
} {
    let newOffsetX: number;
    let newOffsetY: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            newOffsetY = cellRect.y - startPoint.y;
            newOffsetY = newOffsetY > 0 ? minYOffset : newOffsetY;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            newOffsetY = cellRect.y + cellRect.height - startPoint.y;
            newOffsetY = newOffsetY < 0 ? minYOffset : newOffsetY;
            break;
        case 'left':
        case 'right':
            newOffsetY = minYOffset;
            break;
    }

    // Координата X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            newOffsetX = cellRect.x - startPoint.x;
            newOffsetX = newOffsetX > 0 ? minXOffset : newOffsetX;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            newOffsetX = cellRect.x + cellRect.width - startPoint.x;
            newOffsetX = newOffsetX < 0 ? minXOffset : newOffsetX;
            break;
        case 'top':
        case 'bottom':
            newOffsetX = minXOffset;
            break;
    }

    return {
        x: newOffsetX,
        y: newOffsetY,
    };
}

export function getEndPoint(
    startPoint: TPoint,
    direction: TResizerDirection,
    offsetX: number,
    offsetY: number
): TPoint {
    let x: number;
    let y: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            y = startPoint.y - -1 * offsetY + 2;
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            y = startPoint.y + offsetY - 2;
            break;
    }

    // Координата X
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            x = startPoint.x - -1 * offsetX + 2;
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            x = startPoint.x + offsetX - 2;
            break;
    }

    return {
        x,
        y,
    };
}

type TKeyPairs = {
    topLeft: TKeyPair;
    bottomRight: TKeyPair;
};

export function getDragEndParams(
    startKeyPairs: TKeyPairs,
    endKeyPair: TKeyPair,
    direction: TResizerDirection
): TKeyPairs {
    const result: TKeyPairs = {
        topLeft: {
            itemKey: undefined,
            columnKey: undefined,
        },
        bottomRight: {
            itemKey: undefined,
            columnKey: undefined,
        },
    };
    switch (direction) {
        case 'top': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'topLeft': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'topRight': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = endKeyPair.columnKey;
            break;
        }
        case 'bottom': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomLeft': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = endKeyPair.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomRight': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
        case 'left': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'right': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
    }

    return result;
}

export function getRectParams(startPoint: TPoint, offsetX: number, offsetY: number) {
    const transform = (startPosition: number, offset: number) => {
        if (offset < 0) {
            return {
                size: offset * -1,
                position: startPosition + offset,
            };
        }
        return {
            size: offset,
            position: startPosition,
        };
    };

    const xParams = transform(startPoint.x, offsetX);
    const yParams = transform(startPoint.y, offsetY);

    return {
        x: xParams.position,
        y: yParams.position,
        width: xParams.size,
        height: yParams.size,
    };
}

function getViewportBoundingClientRect(element: HTMLElement): TRect {
    // TODO: Сделать меньше хардкода, но это самый производительный вариант.
    //  Т.к. dynamicGrid крайне специфичный, то пока оставляю так.
    element.classList.remove('tw-contents');
    const containerRect = element.getBoundingClientRect();
    element.classList.add('tw-contents');

    const startFixedPart = element.querySelector(`.${START_FIXED_PART_OBSERVER}`);
    const endFixedPart = element.querySelector(`.${END_FIXED_PART_OBSERVER}`);

    if (startFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        const startFixedRectWidth = startFixedPart.getBoundingClientRect().width;
        containerRect.x += startFixedRectWidth;
        containerRect.width -= startFixedRectWidth;
    }

    if (endFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        containerRect.width -= endFixedPart.getBoundingClientRect().width;
    }

    return containerRect;
}

/**
 * Проверяем, что ячейку можно выделить драгом.
 * Ячейка должна быть полностью видимой.
 * @param e React.MouseEvent.
 */
export function canSelectCellByDrag(e: React.MouseEvent): boolean {
    const cellRect = (e.target as HTMLDivElement).getBoundingClientRect();
    const xSizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.width;
    const ySizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.height;
    const availableRect = {
        xStart: cellRect.x + xSizeToRich,
        xEnd: cellRect.x + cellRect.width - xSizeToRich,
        yStart: cellRect.y + ySizeToRich,
        yEnd: cellRect.y + cellRect.height - ySizeToRich,
    };

    return (
        e.pageX >= availableRect.xStart &&
        e.pageX <= availableRect.xEnd &&
        e.pageY >= availableRect.yStart &&
        e.pageY <= availableRect.yEnd
    );
}
