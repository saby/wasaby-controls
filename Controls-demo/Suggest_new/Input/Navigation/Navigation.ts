import { Control, TemplateFunction } from 'UI/Base';
import * as MemorySourceFilter from 'Controls-demo/Utils/MemorySourceFilter';
import * as SearchMemory from 'Controls-demo/Search/SearchMemory';
import { Memory } from 'Types/source';
import { query } from 'Application/Env';
import PortionedSearchMemory from 'Controls-demo/list_new/Searching/DataHelpers/PortionedSearchMemory';
import { _departmentsDataLong } from 'Controls-demo/Suggest_new/DemoHelpers/DataCatalog';
import controlTemplate = require('wml!Controls-demo/Suggest_new/Input/Navigation/Navigation');
import suggestTemplate = require('wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate');

export default class extends Control {
    protected _template: TemplateFunction = controlTemplate;
    protected _suggestTemplate: TemplateFunction = suggestTemplate;
    protected _source: Memory;
    protected _navigation: object;
    protected _portionedNavigation: object;
    protected _maxCountNavigation: object;
    protected _portionedSource: PortionedSearchMemory;

    protected _beforeMount(): void {
        this._source = new SearchMemory({
            keyProperty: 'id',
            data: _departmentsDataLong,
            searchParam: 'title',
            filter: MemorySourceFilter(),
        });
        this._portionedSource = new PortionedSearchMemory({
            direction: 'down',
            fastLoadTimeout: query.get.loadTime ? +query.get.loadTime : undefined,
        });
        this._portionedSource.setFastLoad(true);
        this._portionedNavigation = {
            source: 'position',
            view: 'maxCount',
            sourceConfig: {
                direction: 'forward',
                limit: 4,
                position: 0,
                field: 'key',
            },
            viewConfig: {
                maxCountValue: 12,
            },
        };
        this._navigation = {
            source: 'page',
            view: 'page',
            sourceConfig: {
                pageSize: 12,
                page: 0,
                hasMore: false,
            },
        };
        this._maxCountNavigation = {
            source: 'page',
            view: 'maxCount',
            sourceConfig: {
                pageSize: 4,
                page: 0,
                hasMore: false,
            },
            viewConfig: {
                maxCountValue: 12,
            },
        };
    }

    static _styles: string[] = ['Controls-demo/Suggest_new/Index'];
}
