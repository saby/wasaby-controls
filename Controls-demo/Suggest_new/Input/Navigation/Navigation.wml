<div class="controlsDemo__wrapper controlsDemo__flexRow">
   <div data-qa="controlsDemo-SuggestNewNavigation__pageNavigation" class="controlsDemo__wrapper__horizontal controlsDemo__cell">
      <div class="controls-text-label">Постраничная навигация</div>
      <Controls.SuggestInput
              attr:class="controlsDemo__input demo-SuggestNew__pageNavigation"
              bind:value="_suggestCustomValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_navigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.SuggestInput>
  </div>
   <div data-qa="controlsDemo-SuggestNewNavigation__noSettingsNavigation" class="controlsDemo__wrapper__horizontal controlsDemo__cell">
      <div class="controls-text-label">Без указания настроек навигации</div>
      <Controls.SuggestInput
              attr:class="controlsDemo__input demo-SuggestNew__noSettingsNavigation"
              bind:value="_suggestDefaultValue"
              displayProperty="title"
              searchParam="title"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.SuggestInput>
   </div>
   <div data-qa="controlsDemo-SuggestNewNavigation__loadData" class="controlsDemo__wrapper__horizontal controlsDemo__cell">
      <div class="controls-text-label">Подгружать данные, пока не будет достигут порог</div>
      <Controls.SuggestInput
              attr:class="controlsDemo__input demo-SuggestNew__loadData"
              bind:value="_maxCountValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_maxCountNavigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.SuggestInput>
   </div>
    <div data-qa="controlsDemo-SuggestNewNavigation__portionedNavigation" class="controlsDemo__wrapper__horizontal controlsDemo__cell">
        <div class="controls-text-label">Порционная загрузка</div>
        <Controls.SuggestInput
                attr:class="controlsDemo__input demo-SuggestNew__pageNavigationUp"
                displayProperty="title"
                keyProperty="key"
                searchParam="title"
                navigation="{{_portionedNavigation}}"
                source="{{_portionedSource}}">
            <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate">
                <ws:templateOptions demoClass="demo-SuggestInputList" keyProperty="key"/>
            </ws:suggestTemplate>
        </Controls.SuggestInput>
    </div>
   <div data-qa="controlsDemo-SuggestNewNavigation__pageNavigationUp"  class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_down">
      <div class="controls-text-label">Постраничная навигация (вверх)</div>
      <Controls.SuggestInput
              attr:class="controlsDemo__input demo-SuggestNew__pageNavigationUp"
              bind:value="_suggestCustomValueUp"
              displayProperty="title"
              searchParam="title"
              navigation="{{_navigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Input/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.SuggestInput>
  </div>
</div>
