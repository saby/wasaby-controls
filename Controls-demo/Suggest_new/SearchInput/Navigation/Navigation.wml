<div class="controlsDemo__wrapper controlsDemo__flexRow demo-SearchSuggest__container controlsDemo__mb3">
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left controlsDemo_fixedWidth350">
      <div class="controls-text-label">Постраничная навигация</div>
      <Controls.SuggestSearch
              attr:class="demo-SuggestNewSearch__pageNav controlsDemo__input"
              bind:value="_suggestCustomValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_navigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/SearchInput/Navigation/resources/SuggestTemplate"/>
      </Controls.SuggestSearch>
  </div>
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left controlsDemo_fixedWidth350">
      <div class="controls-text-label">Без указания настроек навигации</div>
      <Controls.SuggestSearch
              attr:class="demo-SuggestNewSearch__noSettingsNav controlsDemo__input"
              bind:value="_suggestDefaultValue"
              displayProperty="title"
              searchParam="title"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/SearchInput/Navigation/resources/SuggestTemplate"/>
      </Controls.SuggestSearch>
   </div>
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left">
      <div class="controls-text-label">Подгружать данные, пока не будет достигут порог</div>
      <Controls.SuggestSearch
              attr:class="demo-SuggestNewSearch__autoLoadNav controlsDemo__input"
              bind:value="_maxCountValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_maxCountNavigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/SearchInput/Navigation/resources/SuggestTemplate"/>
      </Controls.SuggestSearch>
   </div>
</div>
