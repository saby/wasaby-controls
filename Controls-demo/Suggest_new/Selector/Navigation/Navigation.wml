<div class="controlsDemo__wrapper controlsDemo__flexRow">
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left">
      <div class="controls-text-label">Постраничная навигация</div>
      <Controls.ComboboxSuggest
              attr:class="controlsDemo__input demo-SuggestNewSelector__pageNav"
              bind:value="_suggestCustomValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_navigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Selector/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.ComboboxSuggest>
  </div>
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left">
      <div class="controls-text-label">Без указания настроек навигации</div>
      <Controls.ComboboxSuggest
              attr:class="controlsDemo__input demo-SuggestNewSelector__noSettingsNav"
              bind:value="_suggestDefaultValue"
              displayProperty="title"
              searchParam="title"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Selector/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.ComboboxSuggest>
   </div>
   <div class="controlsDemo__wrapper__horizontal controlsDemo__cell ControlsDemo-Suggest_padding-left">
      <div class="controls-text-label">Подгружать данные, пока не будет достигут порог</div>
      <Controls.ComboboxSuggest
              attr:class="controlsDemo__input demo-SuggestNewSelector__autoLoadNav"
              bind:value="_maxCountValue"
              displayProperty="title"
              searchParam="title"
              navigation="{{_maxCountNavigation}}"
              source="{{_source}}">
          <ws:suggestTemplate templateName="wml!Controls-demo/Suggest_new/Selector/Navigation/resources/SuggestTemplate">
              <ws:templateOptions demoClass="demo-SuggestInputList"/>
          </ws:suggestTemplate>
      </Controls.ComboboxSuggest>
   </div>
</div>
