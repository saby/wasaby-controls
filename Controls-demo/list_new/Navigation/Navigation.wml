<div class="controlsDemo__wrapper controlsDemo__flexRow controlsDemo__childMinWidth400">

    <div class="controlsDemo__cell controlsDemo__mr4">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/DigitPaging/Index' }}">
            <a href="{{content.href}}">Постраничная навигация</a>
        </Router.router:Reference>
        <Controls-demo.list_new.Navigation.DigitPaging.Index/>
    </div>

    <div class="controlsDemo__cell controlsDemo__mr4 controlsDemo__maxWidth500">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/HasMorePaging/Index' }}">
            <a href="{{content.href}}">Подгрузка данных по кнопке "Ещё"</a>
        </Router.router:Reference>
        <Controls-demo.list_new.Navigation.HasMorePaging.Index/>
    </div>

    <div class="controlsDemo__cell controlsDemo__mr4 controlsDemo__maxWidth500">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/ScrollPaging/Index' }}">
            <a href="{{content.href}}">Навигация без подсчета страниц (навигация по скроллу, навигация с кнопками прокрутки)</a>
        </Router.router:Reference>
        <Controls-demo.list_new.Navigation.ScrollPaging.Index/>
    </div>

    <div class="controlsDemo__cell controlsDemo__mr4">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/Direction/Index' }}">
            <a href="{{content.href}}">Подгрузка данных по скролу</a>
        </Router.router:Reference>
        <Controls-demo.list_new.Navigation.Direction.Index/>
    </div>

    <div class="controlsDemo__cell controlsDemo__mr4">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/MaxCountValue/Index' }}">
            <a href="{{content.href}}">Подгрузка данных по скролу</a>
        </Router.router:Reference>
        <Controls-demo.list_new.Navigation.MaxCountValue.Index/>
    </div>

    <div class="controlsDemo__cell controlsDemo__mr4">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="app/:app"
                                 app="{{ 'Controls-demo/list_new/Navigation/Paging/Index' }}">
            <a href="{{content.href}}">Подгрузка данных по скролу</a>
        </Router.router:Reference>
    </div>
</div>
