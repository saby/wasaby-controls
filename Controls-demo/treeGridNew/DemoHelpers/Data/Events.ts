import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
