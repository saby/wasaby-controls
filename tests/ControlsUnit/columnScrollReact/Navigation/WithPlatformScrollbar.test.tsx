/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { WasabyEvents } from 'UICore/Events';

import { renderAnyComponent } from '../forTestsOnly/helpers';

import { ScrollbarNavigationComponent } from 'Controls/_columnScrollReact/Navigation/Scrollbar';

describe('ControlsUnit/columnScrollReact/Navigation/WithPlatformScrollbar', () => {
    let container = null;

    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    it('Базовый рендер с платформенным скроллбаром, без моков', () => {
        const { asFragment } = renderAnyComponent(
            <ScrollbarNavigationComponent
                className="testClassName__scrollBar"
                fixedWidth={200}
                contentWidth={900}
                viewPortWidth={400}
                position={20}
                readOnly={false}
                scrollBarValign={'bottom'}
                onPositionChangeCallback={() => {}}
                onDraggingChangeCallback={() => {}}
            />,
            { container }
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
