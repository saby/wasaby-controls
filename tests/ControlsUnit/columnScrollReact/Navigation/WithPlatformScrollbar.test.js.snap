// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsUnit/columnScrollReact/Navigation/WithPlatformScrollbar Базовый рендер с платформенным скроллбаром, без моков 1`] = `
<DocumentFragment>
  <div
    class="tw-w-full controls-ColumnScrollReact__acceleration testClassName__scrollBar"
    data-qa="ColumnScrollReact__navigation__scrollbar"
  >
    <div
      class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_horizontal controls-VScrollbar_horizontal_size-s controls-VScrollbar_horizontal-track_size-s ws-invisible controls-VScrollbar_horizontal-track controls-VScrollbar_horizontal-track"
      data-qa="VScrollbar_horizontal"
      style="top: 0px;"
    >
      <div
        class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-s controls-VScrollbar__thumbWrapper_horizontal controls-padding_left-m controls-padding_right-m"
        style="transform: translateX(0px);"
      >
        <div
          class="controls-VScrollbar__thumb controls-VScrollbar__thumb_unaccented controls-VScrollbar__thumb_size-s controls-VScrollbar__thumb_unaccented controls-VScrollbar__thumb_unaccented_horizontal controls-VScrollbar__thumb_horizontal controls-VScrollbar__thumb_horizontal_size-s "
          data-qa="VScrollbar__thumb"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
